app/controllers/products_controller.rb

```diff
@@ -0,0 +1,103 @@
+class ProductsController < ApplicationController
+  before_action :require_login, only: [:index, :edit, :new]
+  before_action :find_product, only: [:show, :edit, :retire, :activate]
+
+  def index
+  end
+
+  def all_products
+    @products = Product.all
+  end
+
+  def new
+    @product = Product.new
+    @action = "create"
+    @method = :post
+  end
+
+  def create
+    merchant = Merchant.find(params[:merchant_id])
+    @product = merchant.products.new(product_params)
+    if @product.save
+      redirect_to merchant_products_path
+    else
+      render :new
+    end
+  end
+
+  def show
+    @reviews = @product.reviews
+    @review = Review.new
+    @can_review = current_merchant_product?
+    @average = @product.average_review.round
+  end
+
+  def edit
+    @action = "update"
+    @method = :patch
+  end
+
+  def update
+    @product = Product.update(params[:id], product_params)
+    if @product.save
+      redirect_to merchant_products_path
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+   Product.find(params[:id]).destroy
+   redirect_to merchant_products_path(params[:merchant_id])
+  end
+
+  def create_review
+    if current_merchant_product?
+      @review = Review.create(review_params)
+    else
+      flash[:error] = "A merchant cannot review their own product."
+    end
+
+    redirect_to product_path(review_params[:product_id])
+  end
+
+  def retire
+    @product.status = "Retired"
+    @product.save!
+    redirect_to merchant_products_path(@product.merchant_id)
+  end
+
+  def activate
+    @product.status = "Active"
+    @product.save!
+    redirect_to merchant_products_path(@product.merchant_id)
+  end
+
+  private
+
+  def current_merchant_product?
```

Comment:
Comment on the purpose of methods like this one would be helpful to others looking at your code. :heart: comments! Always Be Commenting! :smile: